### PR TITLE
feat(stream): implement group topN

### PR DESCRIPTION
**Pull request by Shmiwy**
Signed-off-by: Shmiwy <wyf000219@126.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

implement local topN
At present, I do not add cache for it, and the efficiency of message delivery is low. I delete all previous messages and then insert messages according to the state table. and Currently, only a single group key is supported

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
#3489

### CODE REVIEW

File: src/common/src/array/stream_chunk.rs [hunk 1]
```diff
@@ -32,7 +32,7 @@ use crate::util::hash_util::finalize_hashers;
 /// but always appear in pairs to represent an update operation.
 /// For example, table source, aggregation and outer join can generate updates by themselves,
 /// while most of the other operators only pass through updates with best effort.
-#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq)]
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
```
Comment (skyzh) [inline src/common/src/array/stream_chunk.rs:35]: I don't think we will need `Hash` for `Op`. Try remove it?

File: src/stream/src/executor/managed_state/top_n/top_n_state_new.rs [hunk 2]
```diff
@@ -117,15 +117,19 @@ impl<S: StateStore> ManagedTopNStateNew<S> {
         TopNStateRow::new(pk_ordered, row)
     }
 
-    #[cfg(test)]
     /// This function will return the rows in the range of [`offset`, `offset` + `limit`),
     pub async fn find_range(
         &self,
+        pk_prefix: Option<&Row>,
```
Comment (skyzh) [inline src/stream/src/executor/managed_state/top_n/top_n_state_new.rs:123]: Please add to the docstring of this function on what's the difference between providing / not providing the pk_prefix.

File: src/stream/src/executor/top_n_local.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+        input: Box<dyn Executor>,
+        order_pairs: Vec<OrderPair>,
+        offset_and_limit: (usize, Option<usize>),
+        pk_indices: PkIndices,
+        store: S,
+        table_id: TableId,
+        total_count: usize,
+        executor_id: u64,
+        key_indices: Vec<usize>,
+        group_by: usize,
```
Comment (skyzh) [inline src/stream/src/executor/top_n_local.rs:44]: Add

```
// TODO: support multiple group keys
```

File: src/stream/src/executor/top_n_local.rs [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+#[async_trait]
+impl<S: StateStore> TopNExecutorBase for InnerLocalTopNExecutorNew<S> {
+    async fn apply_chunk(
+        &mut self,
+        chunk: StreamChunk,
+        epoch: u64,
+    ) -> StreamExecutorResult<StreamChunk> {
+        let mut res_ops = Vec::with_capacity(self.limit.unwrap_or_default());
+        let mut res_rows = Vec::with_capacity(self.limit.unwrap_or_default());
+        let num_limit = self.limit.unwrap();
```
Comment (skyzh) [inline src/stream/src/executor/top_n_local.rs:214]: Should be consistent with above? Or we can disallow limit = None in all places.

**Review by skyzh — COMMENTED** [review]
Looks pretty good at the first glance. I was thinking of directly using the `StateTable` API, but `managed_state` also looks like a good fit here. Will do some detailed investigation tomorrow.

File: src/stream/src/executor/top_n_local.rs (hunk 3 above)
Comment (st1page) [reply src/stream/src/executor/top_n_local.rs:45]: is there any problem supporting multiple group keys in this PR? I think maybe it is ok?

Comment (Shmiwy) [reply src/stream/src/executor/top_n_local.rs:45]: let me try

File: src/stream/src/executor/group_top_n.rs [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+        let mut res_ops = Vec::with_capacity(self.limit.unwrap_or(1024));
+        let mut res_rows = Vec::with_capacity(self.limit.unwrap_or(1024));
+
+        // Record all groups encountered in the input
+        let mut unique_group_keys = HashSet::new();
+
+        // Iterate over all the input, identify which groups they cover.
+        for (op, row_ref) in chunk.rows() {
+            let row = row_ref.to_owned_row();
+            let datum = row[self.group_by].clone();
```
Comment (skyzh) [inline src/stream/src/executor/group_top_n.rs:223]: Can we avoid this clone in the future?

**Review by skyzh — APPROVED** [review]
LGTM! By the way, we should migrate TopN executors to use StateTable. I'll create a separate issue for this, and you can arrange a time to get it fixed.

**Pull request merged**